[DataGrid] Make column header cells focusable

DanailH opened the pull request:
Fixes #1022 

Preview: https://deploy-preview-1289--material-ui-x.netlify.app/components/data-grid/

**A general observation - this was trickier than expected and is more of a "nice to have" feature, but at least navigating the grid with the keyboard is now possible**

This PR aims to provide the ability for users to navigate to the column headers using the keyboard. In addition, focusing on a column header cell and pressing `enter` will trigger the sorting, and pressing `ctrl`/`metaKey` + `enter` will open the 3dot 
column menu and focus it so that pressing the arrows will navigate within the column menu. Pressing `escape` when the 3dot column menu is opened will close it and return the focus to the last focused column header cell.

Things that are missing in this PR:
 - currently, there is a bug that when you navigate with the keyboard to the last rendered column header cell and then try to scroll to the next column header cell the scroll is broken and doesn't calculate the positions correctly.
 - as far as I could work out there is no way to navigate to focus a cell in the grid, meaning that when changing focus using `tab` the focus will move through what is in the toolbar and then be lost - IMO it should focus the first cell or column header (needs to be decided)

To be added:
- [x] tests
- [x] probably it will be a good idea to document the new keyboard interactions

dtassone commented on packages/grid/_modules_/grid/components/columnHeaders/ColumnHeaderMenuIcon.tsx:
```diff
@@ -24,12 +24,7 @@ export function ColumnHeaderMenuIcon(props: ColumnHeaderFilterIconProps) {
-      } else {
-        apiRef!.current.hideColumnMenu();
-      }
+      apiRef!.current.toggleColumnMenu(column.field, columnMenuId, columnMenuButtonId);
```
if you click on another 3 dots icon the menu will blink

DanailH replied:
No, it doesn't (or at least I can't see it). What `toggleColumnMenu ` does is what I removed. I just moved that block of code into a separate method.

dtassone commented on packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderItem.tsx:
```diff
@@ -90,16 +94,33 @@ export const GridColumnHeaderItem = ({
       onMouseOut: publish(GRID_COLUMN_HEADER_OUT),
       onMouseEnter: publish(GRID_COLUMN_HEADER_ENTER),
       onMouseLeave: publish(GRID_COLUMN_HEADER_LEAVE),
+      onKeyDown: publish(GRID_CELL_KEYDOWN),
```
it's not a cell

DanailH replied:
I think we should treat it as one. The event can be renamed but I think we should draw a line between a column header and a cell. They are the same but have different purposes. For example, if one day we need to make the column headers editable that would mean that we would probably need to implement the edit logic again. We should discuss it.

dtassone replied:
Here you need to refactor to `GRID_COLUMN_HEADER_KEYDOWN`

dtassone replied:
Then add a `handleColumnHeaderKeyDown` method in `useGridKeyboard`
then check that if it's arrow down key, `navigateCell`
And if you're on row 0 and arrow up, `navigateHeader`

DanailH replied:
I'm ok with the separate handler and events in the `useGridKeyboard` but why do we need a separate one for the `useGridKeyboardNavigation`. Navigation cells is a generic action that we should be able to select each column header regardless if it has actions or not, at least IMO. It looks that a separate handler will have exactly the same check as in the `navigateCells` and still use `setCellFocus`.

DanailH replied:
updated

dtassone commented on packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderItem.tsx:
```diff
@@ -90,16 +94,33 @@ export const GridColumnHeaderItem = ({
 
+  const publishClick = React.useCallback(
+    (eventName: string) => (event: React.MouseEvent) => {
+      apiRef!.current.setCellFocus({
```
you should treat focus separately
Header and cell are separate as well

dtassone replied:
Need to rename the method `setFocus`

DanailH replied:
I can rename it to `setFocus` but I think it makes more sense to treat column headers and cells the same, after all they are all cells but with just different purposes.

dtassone replied:
I see your point but I would prefer to treat them separately. 
I have thought about that in the past. But I came to the result that it won't scale and the code will be more hacky. As rowIndex===null or some kind of if... It also won't have the same parameters...
So I would like to keep the same approach, but for `ColumnHeader` and then call setFocus

DanailH replied:
Alright, I'll look at it. I'm more concerned about the way I currently used to handle the `rowIndex`. I'm using `null` for the column header cells. It is working and it looks clean but I would like another opinion. The issue is that I still need it and the actual grid rows start from `0` which is ok. I also don't think that the column headers should have `rowIndex` of `0`.

DanailH replied:
Updated

oliviertassinari commented:
> as far as I could work out there is no way to navigate to focus a cell in the grid, meaning that when changing focus using tab the focus will move through what is in the toolbar and then be lost - IMO it should focus the first cell or column header (needs to be decided)

@DanailH Agree, moved to #1291.

dtassone commented:
FYI the focus event is refactored/fixed in #1253

dtassone commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboard.ts:
```diff
@@ -150,7 +155,10 @@ export const useGridKeyboard = (
 
   const handleCellKeyDown = React.useCallback(
     (params: GridCellParams, event: React.KeyboardEvent) => {
```
handleCellKeyDown is just for cell, you should do this in `handleKeyDown`

DanailH replied:
updated

dtassone commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboard.ts:
```diff
@@ -92,12 +100,9 @@ export const useGridKeyboard = (
-            ? selectionFromRowIndex
-            : nextCellIndexes.rowIndex,
-        )
+      const nextRowIndex = nextCellIndexes.rowIndex === null ? 0 : params.rowIndex;
```
This should not change as you should never select headers...
Please refactor, to separate the concern for headers...

DanailH replied:
But why, I was left with the impression that you should be able to "select" headers. Basically, if you are on a cell with coordinates 0:0 pressing the up arrow should select the header above it.

DanailH replied:
reverted

dtassone commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboard.ts:
```diff
@@ -150,7 +155,10 @@ export const useGridKeyboard = (
     (params: GridCellParams, event: React.KeyboardEvent) => {
-      if (!isGridCellRoot(document.activeElement)) {
+      if (
+        !isGridCellRoot(document.activeElement) &&
```
that checks should go, if it's an header then it should go in another direction

dtassone commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:
```diff
@@ -66,58 +80,56 @@ export const useGridKeyboardNavigation = (
-
-      const currentColIndex = Number(cellEl.getAttribute('aria-colindex'));
-      const currentRowIndex = Number(cellEl.getAttribute('data-rowindex'));
+      const rowIndex =
```
use params here. `params.rowIndex`

DanailH replied:
It should be as it is because when applying a sorting the `rowIndex` is no longer the one you expect so the focus will start jumping between cells unpredictably.

dtassone commented on packages/grid/_modules_/grid/hooks/features/columnMenu/useGridColumnMenu.ts:
```diff
@@ -9,7 +9,7 @@ export const useGridColumnMenu = (apiRef: GridApiRef): void => {
 
   const showColumnMenu = React.useCallback(
-    (field: string, id: string, labelledby: string) => {
+    (field: string, id?: string, labelledby?: string) => {
```
I maintain my point from my previous PR, this `labelledBy` should not go here.
and type should be `showColumnMenu(id: GridRowId, field: string)`
Api methods on apiRef should just take the min params

DanailH replied:
I agree we need a better solution for this. I did it so it doesn't block the development of this feature but still supports the other one. If possible I would prefer to ignore it for now and tackle that problem separetly.

DanailH replied:
created -> https://github.com/mui-org/material-ui-x/issues/1365

dtassone commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:
```diff
@@ -28,17 +28,31 @@ const getNextCellIndexes = (code: string, indexes: GridCellIndexCoordinates) =>
+      if (indexes.rowIndex === null) {
+        return { rowIndex: null, colIndex: indexes.colIndex - 1 };
+      }
+      return { ...indexes, colIndex: indexes.colIndex - 1 };
```
no need refactoring here. if there is no rowIndex then leave it

DanailH replied:
If there is no `rowIndex` then it will be `undefined` which is a problem. I think it should still have a value just be `null`.

DanailH replied:
Updated

dtassone commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:
```diff
@@ -28,17 +28,31 @@ const getNextCellIndexes = (code: string, indexes: GridCellIndexCoordinates) =>
+        return { ...indexes, rowIndex: null };
+      }
+
+      return { ...indexes, rowIndex: indexes.rowIndex - 1 };
```
return { ...indexes, rowIndex:  indexes.rowIndex ? indexes.rowIndex - 1 : 0};

DanailH replied:
It can't be just 0 because 0 is the first row. So it needs to be able to identify a header somehow.

DanailH replied:
Updated

dtassone commented:
FYI more about events  in #1301

dtassone commented:
This PR should also cover #1295, and #1291
As they are closely related issues. Is that ok?

oliviertassinari commented:
Please, let's always try to break pull requests into chunks **as small as possible**, while we kept delivering value to the users, and keep the PR overhead small. A possible breakdown:

1. Make the header cell accessible in the arrow navigation. WAI-ARIA mentions that headers shouldn't be focusable if there are no actions on them, but since we always have a column menu, maybe we can ignore this aspect. Make sort and opening column menu keyboard accessible when a header cell is focused 
3. Fix #1291: Removing the tabIndex from the root element, replacing it with the first cell being tabbable. 
4. Fix <kbd>Tab</kbd> Moving to the next cell when in read mode, it should move to the next widget. 
5. Fix #1295: Seems completely off-topic.

oliviertassinari commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:
```diff
@@ -28,17 +28,31 @@ const getNextCellIndexes = (code: string, indexes: GridCellIndexCoordinates) =>
-  }
-  if (code === 'ArrowUp') {
-    return { ...indexes, rowIndex: indexes.rowIndex - 1 };
+  switch (code) {
```
code is confusing, it's coming from `event.key` 

```suggestion
  switch (key) {
```

dtassone commented:
> Please, let's always try to break pull requests into chunks **as small as possible**, while we kept delivering value to the users, in each chunk. Of course, it only makes sense if the PR overhead stays small (e.g. 1 line changes have a high PR overhead). A possible breakdown:
> 
> 1. Fix #1022 Make the header cell accessible in the arrow navigation. WAI-ARIA mentions that headers shouldn't be focusable if there are no actions on them, but since we always have a column menu, maybe we can ignore this aspect. Make sort and opening column menu keyboard accessible when a header cell is focused

- If the column is not sortable and we disable the column menu then, there is no button left in the col so it should not be tab-able.

> 2. Fix #1291: Removing the tabIndex from the root element, replacing it with the first cell being tabbable.

This should be the same fix as above. We should aim for a global solution for this issue and not only focus on the little issue.
A solution could be 

``` tsx
function useGridKeyboard(apiRef){
....
const handleTabKeyDown= (params: ?? CellParams | ColumnHeaderParams, event: KeyboardEvent)=> {
   //Maybe something like
   const tabable = apiRef.current.getNextTabableParams(params)
    if(tabable.type === 'cell') {
       `apiRef.current.setCellFocus(id, field) `
    }
    if(tabable.type === 'columnHeader') {
       `apiRef.current.setColumnHeaderFocus(field) ` => should call scroll to header
    } 
else {
       `apiRef.current.setElementFocus(tabable.element) `
    }
}

useApiEventHandler('tabKeyDown', handleTabKeyDown);
}
// useTabable  ??
/*
const getNextTabableParams = ?? 
```

Tricky bit here 
1. Check if preference is open? Column menu open?? Should it be handled already?
2. add data-tabable to all tabable element? issue with virtualisation => Not IDEAL
3. add a tabable param in GridCellParams and in GridColumnHeader
  - For column header 
       -  Create  `apiRef.current.getNextTabableColumn(currentField): GridColumnHeaderParams`
         - You can get the column after the current index and go next until you get the next tabable one.
         - Or you can get all columns and find the next tabable by filetering by not sortable and disableColumnMenu... 
       -  Then get its params and apply focus using `apiRef.current.setColumnHeaderFocus(field)`
  - For cell 
     - if we focus on all cell using tab regardless of its content then it working with the current solution

> 3. Fix Tab Moving to the next cell when in read mode, it should move to the next widget.

Not sure I understand this one.

> 4. Fix #1295

This the part 1. of my solution above

While you're there, and you know everything about the area. You might as well fix it. It is closely related. If somebody else picks it up, there would be learning curve, and it create new issues...
It can be another PR, but one big PR that provide a global solution for accessibility is great, as you thought of the big picture. Then we can have small ones for little issues regarding edge cases

oliviertassinari commented:
> If the column is not sortable and we disable the column menu then, there is no button left in the col so it should not be tabbable

@dtassone With the roving-index, it would only be tabbable if the arrow navigation was used to focus the cell. I believe you meant **focusable**, not tabbable.

> Not sure I understand this one.

I mean this bug: https://github.com/mui-org/material-ui-x/issues/1291#issuecomment-807086470

> I believe this is for the toolbar, footer and navigating in and outisde the grid

As far as I have looked at #1295, it's a regression from the latest release, which I can narrow down to this commit: https://github.com/mui-org/material-ui-x/pull/1205/commits/5169c949c17e53f299ecb8af13c240bd7520776e

DanailH commented:
Alright, to summerise:
- I will refactor the logic to separate the concerns between column header cells and normal cells (for now the logic will be mostly duplicated but in time it may diverge).
- Regarding the comment about including additional fixes in this PR - I agree that it should be on point. In its current state, the only problem that this PR solves is this one https://github.com/mui-org/material-ui-x/issues/1022. The rest should be done in separate PRs.
- I need to extend the solution a but because currently, every cell is focusable (to be honest I think it should be like this because otherwise, it may create unnecessary complications. After all each cell is focusable and not all cells have actions).
- Last but not least the problem with the scroll needs to be resolved because otherwise, the feature is buggy, meaning that when you use the keyboard to move from column header to column header when trying to focus the next cell that is out of the viewport the positions are not aligned.

oliviertassinari commented:
> currently, every cell is focusable

👍 For keeping this behavior and ignoring the potential need for non focusable cells.

dtassone commented on packages/grid/_modules_/grid/hooks/features/virtualization/useGridVirtualRows.ts:
```diff
@@ -221,9 +221,8 @@ export const useGridVirtualRows = (
-      const currentRowIndex = params.rowIndex === null ? 0 : params.rowIndex;
       const currentRowPage =
-        (currentRowIndex - gridState.pagination.page * gridState.pagination.pageSize) /
+        (params.rowIndex - gridState.pagination.page * gridState.pagination.pageSize) /
```
I had to refactor this as well

DanailH commented:
> Much better approach 👍

Yes, it looks cleaner. Once you merge the PR with the render fixes I'll rebase and see if this will fix the scroll problem. In the meantime, I'll add tests and docs.

DanailH commented:
I've added tests and updated the docs.

The bug related to scrolling left is still present. I would need this PR https://github.com/mui-org/material-ui-x/pull/1319 to see if it will potentially fix it (we've discussed it with @dtassone)

oliviertassinari reviewed (changes requested):
I have noticed the following issues with the UX:

1. As soon as there is a sort in place, the arrow navigation is no longer working
![sorting](https://user-images.githubusercontent.com/3165635/113156300-24e9e980-923a-11eb-9263-679d0ed85e94.gif)
2. Open https://deploy-preview-1289--material-ui-x.netlify.app/components/data-grid/, scroll to the bottom. The arrow navigation is so slow you can't use it.

dtassone commented:
As mention, this PR might have different behaviour when it is rebased with #1319.

oliviertassinari commented on packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx:
```diff
@@ -189,5 +202,40 @@ describe('<DataGrid /> - Keyboard', () => {
+    getColumnHeaderCell(1).focus();
+    expect(getActiveCell()).to.equal('null-1');
+    expect(getColumnValues(1)).to.deep.equal(['John', 'Doe']);
+    fireEvent.keyDown(document.activeElement!, ENTER_KEY);
+    fireEvent.keyDown(document.activeElement!, ENTER_KEY);
```
The cost is the indirection. Having to carry a variable around. We don't aim for DRY in the tests.

oliviertassinari replied:
We could open an RFC to discuss this, to change the current standard.

DanailH replied:
I don't really have a preference - I followed the same approach as the others were written. Personally having the object directly for me is better as it saves me the mental overhead of having to check what the object is equal to.

oliviertassinari replied:
I ran a search in the codebase, we have a direct object in 327 cases. We have an abstract object in 3 cases.

DanailH replied:
updated

DanailH commented on docs/src/pages/components/data-grid/accessibility/accessibility.md:
```diff
@@ -52,6 +52,8 @@ Use the arrow keys to move the focus.
 |                                 <kbd>CTRL</kbd> + <kbd>A</kbd> | Select all rows                                   |
 |                                 <kbd>CTRL</kbd> + <kbd>C</kbd> | Copy the currently selected row                   |
 |                                <kbd>CTRL</kbd> + Click on cell | Enable multi-selection                            |
+|                                               <kbd>ENTER</kbd> | Sort column when column header is selected        |
+|                             <kbd>CTRL</kbd> + <kbd>ENTER</kbd> | Open column menu when column header is selected   |
```
Noted

DanailH commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:
```diff
@@ -135,15 +162,76 @@ export const useGridKeyboardNavigation = (
     (nextCellIndexes: GridCellIndexCoordinates) => {
-      apiRef.current.scrollToIndexes(nextCellIndexes);
+      if (nextCellIndexes !== null) {
+        apiRef.current.scrollToIndexes(nextCellIndexes);
```
That one is no longer needed. I'll remove it.

DanailH commented on packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderItem.tsx:
```diff
@@ -143,8 +161,16 @@ export const GridColumnHeaderItem = ({
 
+  React.useLayoutEffect(() => {
+    const lastMenuState = apiRef!.current.getState().columnMenu;
+    if (hasFocus && headerCellRef.current && !lastMenuState.open) {
```
Updated

DanailH commented:
> I have noticed the following issues with the UX:
> 
> 1. As soon as there is a sort in place, the arrow navigation is no longer working
>    ![sorting](https://user-images.githubusercontent.com/3165635/113156300-24e9e980-923a-11eb-9263-679d0ed85e94.gif)
> 2. Open https://deploy-preview-1289--material-ui-x.netlify.app/components/data-grid/, scroll to the bottom. The arrow navigation is so slow a end-user can't use it.

Fixed. Actually, the only problem that persists is when you are focusing on a column header and you try to select the next one that is not rendered then the scroll is misaligned.
If you are focusing on a cell and then scroll the grid by selecting the cells not yet rendered everything works as expected.

oliviertassinari commented on packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderItem.tsx:
```diff
@@ -135,8 +142,15 @@ export const GridColumnHeaderItem = ({
   );
   const columnMenuIconButton = <ColumnHeaderMenuIcon column={column} />;
 
+  React.useLayoutEffect(() => {
+    if (hasFocus && apiRef!.current.getState().columnMenu) {
+      headerCellRef.current!.focus();
+    }
+  });
```
I don't get why we force the focus at each render. This seems to be the same problem as with the cell. I would be surprised if this is ever correct. For instance, could it be why the column menu isn't usable with the keyboard?

I even wonder if this focus state should ever be in the state. The focus is typically a state you might always prefer to manage imperatively.

DanailH replied:
Am not at each render. This is to check which cell is currently the one that should be focused on and not steal the focus when the column menu is opened. The cell that is to be focused is stored in the state.

oliviertassinari reviewed (changes requested):
- When pressing <kbd>Space</kbd> on a header cell:

<img width="388" alt="Screenshot 2021-04-02 at 01 34 02" src="https://user-images.githubusercontent.com/3165635/113364344-86ea4200-9353-11eb-84f1-119a5a7c20dc.png">

- The column menu is not keyboard accessible. I suspect the header cell steal the focus.

oliviertassinari commented:
Damien mentions that 876633bef66de6fab1679d09475bb333f0bf8052 could interfer. Shouldn't we *git rebase*?

DanailH commented:
> Damien mentioned that [876633b](https://github.com/mui-org/material-ui-x/commit/876633bef66de6fab1679d09475bb333f0bf8052) could interfer. Shouldn't we _git rebase_? The current base is [728bad3](https://github.com/mui-org/material-ui-x/commit/728bad3ef1f481a84a042d9bb44da970b56b5c87) (one commit before)

I pulled the latest yesterday. I'll do it again.

DanailH commented:
> * When pressing Space on a header cell:
> 
> <img alt="Screenshot 2021-04-02 at 01 34 02" width="388" src="https://user-images.githubusercontent.com/3165635/113364344-86ea4200-9353-11eb-84f1-119a5a7c20dc.png">
> 
> * The column menu is not keyboard accessible. I suspect the header cell steal the focus.

Fixed. The question is should we `preventDefault` when space is pressed. Currently, I don't.

I also check - I do have the latest changes from `master` (the scroll is fixed)

oliviertassinari commented on docs/src/pages/components/data-grid/accessibility/accessibility.md:
```diff
@@ -28,20 +28,20 @@ The grid responds to keyboard interactions from the user as well as emitting eve
-|              <kbd>Page Down</kbd> | Navigate to the previous scrollable page      |
+|                              Keys | Description                                      |
+| --------------------------------: | :----------------------------------------------- |
+|                    <kbd>Tab</kbd> | Navigate between selectable elements             |
```
<kbd>Tab</kbd> should move to the next tabbable element, but instead, it moves to the next focusable one. As far as I know, it's wrong.

```suggestion
```

DanailH replied:
I'll handle that one in here #1327

oliviertassinari reviewed (changes requested):
> I do have the latest changes from master

My bad, I was expecting a linear git history. But instead:

<img width="679" alt="Screenshot 2021-04-02 at 11 29 35" src="https://user-images.githubusercontent.com/3165635/113403343-c5f5b300-93a6-11eb-88d7-24e1ec9d0b03.png">

I have no idea what's based on. I would personally always rebase and keep the history linear, simply to be able to reason with it.

---

Regarding what's left, I believe we have:

> The question is should we preventDefault when space is pressed. Currently, I don't.

- [ ] Yes, I think so. `<button>` triggers on <kbd>Space</kbd> and <kbd>Enter</kbd>. In https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html, space and enter trigger the sort. As far as I know, we should handle space, trigger the sort, and prevent default. ag-Grid does it wrong.
- [ ] <kbd>Tab</kbd> should move to the next tabbable element, but instead, it moves to the next focusable one. As far as I know, it's wrong. ag-Grid does it wrong. https://www.w3.org/TR/wai-aria-practices/#grid
- [ ] When focusing a header cell <kbd>Page down</kbd> should move the cell focus, not scroll. Same for <kbd>Page up</kbd>

DanailH commented:
From the "What is left list":
1. I'll handle it here.
2. The part related to the `tab` navigation will be handled here https://github.com/mui-org/material-ui-x/pull/1327
3. Regarding the last one - pressing `Page Down` or `Page Up` which cell should be focused on, the last one in the column? Also if you are in the column header and press `Page Up` should you focus on the first cell of the column? To me, it seems but counterintuitive to have actions mapped to `Page Down` or `Page Up` when a column header is selected.

oliviertassinari commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboard.ts:
```diff
@@ -212,7 +212,8 @@ export const useGridKeyboard = (
 
-      if (isEnterKey(event.key)) {
+      if (isEnterKey(event.key) || isSpaceKey(event.key)) {
+        event.preventDefault();
```
We should only prevent default if space, enter doesn't need it.

DanailH replied:
fixed

oliviertassinari commented:
> Regarding the last one - pressing Page Down or Page Up which cell should be focused on, the last one in the column? Also if you are in the column header and press Page Up should you focus on the first cell of the column? To me, it seems but counterintuitive to have actions mapped to Page Down or Page Up when a column header is selected.

The same one as if we were on a cell.

DanailH commented:
@oliviertassinari I updated the behaviour for the `Page Up` and `Page Down` keys. Now they work as expected.

m4theushw commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:
```diff
@@ -137,13 +164,78 @@ export const useGridKeyboardNavigation = (
+        if (key.indexOf('Down') > -1) {
+          apiRef.current.setCellFocus({
+            colIndex: params.colIndex,
+            rowIndex: containerSizes!.viewportPageSize - 1,
```
Why move the focus back to the last but one visible row? Should not be the row that had focus before <kbd>PageUp</kbd> or fallback to the first?

DanailH replied:
Because that is what the bottom of the page is. An interaction like this is not really logical at least IMO.

m4theushw replied:
I asked because I found this a bit weird.

https://user-images.githubusercontent.com/42154031/113713326-93f19300-96bd-11eb-9a07-31821218af79.mp4

m4theushw commented on packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx:
```diff
@@ -151,6 +157,12 @@ describe('<DataGrid /> - Keyboard', () => {
     fireEvent.keyDown(document.activeElement!, { key: 'ArrowUp' });
     expect(getActiveCell()).to.equal('0-0');
+    fireEvent.keyDown(document.activeElement!, { key: 'ArrowUp' });
+    expect(getActiveCell()).to.equal('null-1');
```
Add `getActiveHeaderCell` to check focus on header cells?

DanailH replied:
The thing is that cells do not have `rowIndex`, that's why there is a `null`. Adding one would only benefit the test as it isn't used anywhere else in the logic.

Regarding, the `getActiveHeaderCell` -> it can be a possible solution and it will only return the `colIndex` so basically,
`getActiveHeaderCell() === 'n' \\ where n is the value of the colIndex`

DanailH replied:
Updated

m4theushw commented on packages/grid/_modules_/grid/constants/eventsConstants.ts:
```diff
@@ -39,6 +39,9 @@ export const GRID_ROW_LEAVE = 'rowLeave';
 export const GRID_ROW_SELECTED = 'rowSelected';
 
+export const GRID_COLUMN_HEADER_FOCUS = 'columnHeaderFocus';
+export const GRID_COLUMN_HEADER_NAVIGATION_KEYDOWN = 'columnHeaderNavigationKeyDown';
```
GRID_COLUMN_HEADER_NAVIGATION_KEY_DOWN?

m4theushw replied:
Based on this line of thinking, other constants should be changed as well:

```diff
- export const GRID_CELL_DOUBLE_CLICK = 'cellDoubleClick';
+ export const GRID_CELL_DBLCLICK = 'cellDblclick';
```

We have to agree on a single naming strategy and stick to it. Currently, the constants file has different patterns. My preference is for `GRID_COLUMN_HEADER_NAVIGATION_KEYDOWN` as I think that each spoken word should be preceded by an underscore (except for the first word). However, `GRID_COLUMN_HEADER_NAVIGATION_KEYDOWN` is also fine if we make it clear that the DOM event must be used as suffix _as it is_.

Obviously, updating all constants is out of scope of this PR. I think this is something https://github.com/mui-org/material-ui-x/issues/1301 should address too.

oliviertassinari replied:
+1 for: 

- consistency
- outside of the scope of this PR
- opening an RFC with a proposal
- present tense (this is an event)
- staying close to the event name (e.g. `document.addEventListener('dbclick')` for the dom, but prop in React is `onDoubleClick` so 🤷‍♂️)

DanailH replied:
In that case I would push the change `GRID_COLUMN_HEADER_NAVIGATION_KEY_DOWN` into a separate one as well.

oliviertassinari replied:
I think that there are two levels to this discussion. The first one is about having the string match the format of the const. Here, it's a new line, and should match.

Then, there is a larger discussion about the whole naming consistent of the costs, which can happen with an RFC.

DanailH replied:
Agree, I went with this one `export const GRID_COLUMN_HEADER_NAVIGATION_KEYDOWN = 'columnHeaderNavigationKeydown';`

m4theushw commented on packages/grid/_modules_/grid/constants/eventsConstants.ts:
```diff
@@ -39,6 +39,9 @@ export const GRID_ROW_LEAVE = 'rowLeave';
 
+export const GRID_COLUMN_HEADER_FOCUS = 'columnHeaderFocus';
+export const GRID_COLUMN_HEADER_NAVIGATION_KEYDOWN = 'columnHeaderNavigationKeyDown';
+export const GRID_COLUMN_HEADER_KEYDOWN = 'columnHeaderKeyDown';
```
GRID_COLUMN_HEADER_KEY_DOWN?

DanailH replied:
Updated

m4theushw reviewed (commented):
Would be nice to have a test case with <kbd>PageUp</kbd> and <kbd>PageDown</kbd>.

oliviertassinari commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:
```diff
@@ -137,13 +164,78 @@ export const useGridKeyboardNavigation = (
+        return;
+      }
+
+      nextColumnHeaderIndexes!.colIndex =
+        nextColumnHeaderIndexes!.colIndex <= 0 ? 0 : nextColumnHeaderIndexes!.colIndex;
```
Seems simpler to read and to identify patterns later on (clamp):

```suggestion
      nextColumnHeaderIndexes!.colIndex = Math.min(0, nextColumnHeaderIndexes!.colIndex);
```

x Math.max

oliviertassinari replied:
https://github.com/mui-org/material-ui/blob/bcb61849635b44c44c687a11e12a96087d582397/packages/material-ui-unstyled/src/SliderUnstyled/SliderUnstyled.js#L25

DanailH replied:
To have it working like this it will look more like that:
`nextColumnHeaderIndexes!.colIndex = Math.max(0, nextColumnHeaderIndexes!.colIndex);`

DanailH commented on packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx:
```diff
@@ -189,5 +201,40 @@ describe('<DataGrid /> - Keyboard', () => {
+    expect(getActiveCell()).to.equal('null-1');
+    expect(getColumnValues(1)).to.deep.equal(['John', 'Doe']);
+    fireEvent.keyDown(document.activeElement!, { key: 'Enter' });
+    fireEvent.keyDown(document.activeElement!, { key: 'Enter' });
```
The first one doesn't sort the cell in a descending way (it's the same with clicks).

DanailH commented on docs/src/pages/components/data-grid/accessibility/accessibility.md:
```diff
@@ -52,6 +52,8 @@ Use the arrow keys to move the focus.
 |                                 <kbd>CTRL</kbd> + <kbd>A</kbd> | Select all rows                                   |
 |                                 <kbd>CTRL</kbd> + <kbd>C</kbd> | Copy the currently selected row                   |
 |                                <kbd>CTRL</kbd> + Click on cell | Enable multi-selection                            |
+|                                               <kbd>Enter</kbd> | Sort column when column header is selected        |
+|                             <kbd>CTRL</kbd> + <kbd>Enter</kbd> | Open column menu when column header is selected   |
```
Maybe `selected` is not the best word to describe it, `focused` sounds better?

DanailH replied:
Updated

DanailH commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboard.ts:
```diff
@@ -187,8 +195,43 @@ export const useGridKeyboard = (
+        return;
+      }
+
+      if (isSpaceKey(event.key)) {
```
I understand, and I have no problems changing it but in favour of consistency, it's better to use that helper. If we agree that that would be the way forward I suggest to change it everywhere and drop the helpers.

DanailH commented on packages/grid/_modules_/grid/hooks/features/columnMenu/useGridColumnMenu.ts:
```diff
@@ -30,6 +30,19 @@ export const useGridColumnMenu = (apiRef: GridApiRef): void => {
+    (field: string, id?: string, labelledby?: string) => {
+      logger.debug('Toggle Column Menu');
+      const lastMenuState = gridState.columnMenu;
+      if (!lastMenuState.open || lastMenuState.field !== field) {
```
Done

DanailH commented:
> Would be nice to have a test case with PageUp and PageDown.

Added one additional test.

dtassone commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:
```diff
@@ -66,59 +92,60 @@ export const useGridKeyboardNavigation = (
     (params: GridCellParams, event: React.KeyboardEvent) => {
       event.preventDefault();
-
+      const { colIndex } = params;
```
you can get the rowIndex here as well. Otherwise you have to update the params type

DanailH replied:
I answered that before. 

The `rowIndex` that comes from the params changes depending on the sorting. Getting it from the params messes up the keyboard navigation if sorting is applied.

dtassone replied:
Yes there was an issue with the rowIndex when a column was sorted. It is fixed now

DanailH replied:
The last time I tried it didn't work. I'll try it again.

dtassone commented on packages/grid/_modules_/grid/hooks/features/keyboard/useGridKeyboardNavigation.ts:
```diff
@@ -66,59 +92,60 @@ export const useGridKeyboardNavigation = (
+
       nextCellIndexes.rowIndex =
-        nextCellIndexes.rowIndex >= rowCount && rowCount > 0
+        nextCellIndexes.rowIndex !== null && nextCellIndexes.rowIndex >= rowCount && rowCount > 0
```
in what case can it be null?

DanailH replied:
That one is not needed, it was a check from before. Removed

dtassone commented on packages/grid/_modules_/grid/models/api/gridVirtualizationApi.ts:
```diff
@@ -17,7 +18,7 @@ export interface GridVirtualizationApi {
    * @returns boolean Return if the index was outside of the viewport and the grid has to scroll to reach the target.
    */
-  scrollToIndexes: (params: GridCellIndexCoordinates) => boolean;
+  scrollToIndexes: (params: Optional<GridCellIndexCoordinates, 'rowIndex'>) => boolean;
```
A bit difficult to get why rowIndex becomes optional. Maybe we can add GridColumnHeaderIndexCoordinates? 🤔

DanailH replied:
It because option so it can `scrollToIndexes` can handle horizontal scrolling only.

dtassone replied:
yes in the context of the column header. So the type should be `GridCellIndexCoordinates | GridColumnHeaderIndexCoordinates`

DanailH replied:
Yes, it's either that or what you said. The problem with the `GridCellIndexCoordinates | GridColumnHeaderIndexCoordinates` type is that the type needs to be explicit everywhere `rowIndex` is used, so `param.rowIndex as GridCellIndexCoordinates`.

DanailH commented:
This now contains the changes from part 1 and can have a final review.

oliviertassinari reviewed (approved):
It looks good enough for this iteration. Two follows-up I can identify (on top of adding tests for the <kbd>Tab</kbd>, but related):

- [ ] <kbd>Space</kbd> doesn't trigger the checkbox selection all action when the focus is on the "select all" column header cell. I'm not sure about <kbd>Enter</kbd> as a native checkbox doesn't check with such action.
- [ ] The checkboxes shouldn't be tabbable. I believe we should add`tabIndex={-1}`. ag-Grid does it, it's also what the a11y guidelines documents, that we remove cell's content from the tab order.
<img width="194" alt="Screenshot 2021-04-12 at 20 27 13" src="https://user-images.githubusercontent.com/3165635/114443004-7d6f9e00-9bcd-11eb-81be-46fd7a40c65d.png">

DanailH commented:
> It looks good enough for this iteration. Two follows-up I can identify (on top of adding tests for the Tab, but related):
> 
> * [ ]  Space doesn't trigger the checkbox selection all action when the focus is on the "select all" column header cell. I'm not sure about Enter as a native checkbox doesn't check with such action.
> * [ ]  The checkboxes shouldn't be tabbable. I believe we should add`tabIndex={-1}`. ag-Grid does it, it's also what the a11y guidelines documents, that we remove cell's content from the tab order.
> 
> <img alt="Screenshot 2021-04-12 at 20 27 13" width="194" src="https://user-images.githubusercontent.com/3165635/114443004-7d6f9e00-9bcd-11eb-81be-46fd7a40c65d.png">

@oliviertassinari thanks for the final checks. I removed the `tabIndex` from the checkboxes and from the buttons in the column header because they have dedicated keyboard interactions.

For the first one - I think we should also handle the individual row selection. Using `Enter` will probably be better because `Space` has another interaction and having `Space` for the column headers and a different one for the row selection will be inconsistent.

oliviertassinari commented:
>  I removed the tabIndex from the checkboxes and from the buttons in the column header because they have dedicated keyboard interactions.

@DanailH Yes, this sounds good. However, you forgot about the checkbox that is in the column header. I could instantly spot it by trying to use the keyboard only to navigate the grid.

<img width="197" alt="Screenshot 2021-04-13 at 12 07 00" src="https://user-images.githubusercontent.com/3165635/114535851-c3ba1100-9c50-11eb-9e9f-137b5e3b0ccc.png">

> Using Enter will probably be better because Space has another interaction

Note that this column is not sortable, enter and space currently do nothing.

ramyab03 commented:
Hi Team,
                 We are using MUI DataGrid @mui/x-data-grid in our React application. 

We are not using Sorting, filtering, ColumnMenu or ColumnSelector features on the DataGrid header. So the headers are just plain text. In this case we don't want users to be able click into (mouse) or tab into (keyboard) the header row - since there is no action that occurs when they click on the headers. 

Is there any way that this can be achieved?
